pymove/utils/db.py

```diff
@@ -0,0 +1,439 @@
+    dataframe,
+    dbname="postgres",
+    user="postgres",
+    password="",
+    host="localhost",
+    port=5432,
+):
+    """
+    Saves a dataframe to a postgres table.
+
+    Parameters
+    ----------
+    table : string
+        Name of the table
+
+    dataframe : dataframe object
+        The dataframe to be saved
+
+    dbname : string, default 'postgres'
+        Name of the database
+
+    user : string, default 'postgres'
+        The user connecting to the database
+
+    password : string, default ''
+        The password of the database
+
+    host : string, default 'localhost'
+        The address of the database
+
+    port : int, default 5432
+        The port of the database in the host
+    """
+    cols = dataframe.columns
+    columns = ",".join(cols)
+    values = ",".join(["%s"] * len(cols))
+
+    conn = None
+    sql = "INSERT INTO %s(%s) VALUES(%s)" % (table, columns, values)
+    try:
+        conn = connect_postgres(dbname, user, password, host, port)
+        cur = conn.cursor()
+        cur.execute("DELETE FROM %s" % (table))
+        cur.executemany(sql, dataframe.values)
+        conn.commit()
+        cur.close()
+    except Exception as e:
+        if conn is not None:
+            conn.close()
+        raise e
+    finally:
+        if conn is not None:
+            conn.close()
+
+
+def read_postgres(
+    query,
+    in_memory=True,
+    type_="pandas",
+    dbname="postgres",
+    user="postgres",
+    password="",
+    host="localhost",
+    port=5432,
+):
+    """
+    Builds a dataframe from a query to a postgres database.
+
+    Parameters
+    ----------
+    query : string
+        Sql query
+
+    in_memory : bool, default True
+        Whether te operation will be executed in memory
+
+    type_ : 'pandas', 'dask' or None, defaults 'pandas'
+        It will try to convert the dataframe into a MoveDataFrame
+
+    dbname : string, default 'postgres'
+        Name of the database
+
+    user : string, default 'postgres'
+        The user connecting to the database
+
+    password : string, default ''
+        The password of the database
+
+    host : string, default 'localhost'
+        The address of the database
+
+    port : int, default 5432
+        The port of the database in the host
+
+    Returns
+    -------
+    dataframe: PandasMoveDataFrame, DaskMoveDataFrame or PandasDataFrame
+        a dataframe object
+    """
+    conn = None
+    dataframe = None
+    try:
+        conn = connect_postgres(dbname, user, password, host, port)
+        if in_memory:
+            dataframe = read_sql_inmem_uncompressed(query, conn)
+        else:
+            dataframe = read_sql_tmpfile(query, conn)
+    except Exception as e:
+        if conn is not None:
+            conn.close()
+        raise e
+    finally:
+        if conn is not None:
+            conn.close()
+        try:
+            return MoveDataFrame(dataframe, type_=type_)
+        except Exception:
+            return dataframe
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [return statement in finally block may swallow exception](https://app.codacy.com/gh/InsightLab/PyMove/pullRequest?prid=5535892)